nengo_spa/tests/test_semantic_pointer.py

```diff
@@ -317,9 +318,136 @@ def test_name():
+        # that is because sqrt is a multi-valued function
+        # assert np.allclose(((x**a)**b).v, (x**(a*b)).v)
+
+
+@pytest.mark.parametrize("d", [1, 4, 9, 16, 129, 512])
+def test_fractional_binding_unitary(d, rng):
+    # tests a number of properties of fractional binding with nondegenerate
+    # and unitary vectors
+
+    gen = UnitaryVectors(d, algebra=HrrAlgebra(), rng=rng)
+    for _ in range(100):
+        x = SemanticPointer(next(gen)).nondegenerate()
+
+        # check that making it nondegenerate or unitary again
+        # does not change the vector
+        assert np.allclose((x.nondegenerate()).v, x.v)
+        assert np.allclose((x.unitary()).v, x.v)
+
+        # 1/x is the inverse (since unitary)
+        assert np.allclose((x ** (-1)).v, (~x).v)
+
+        # x**0 gives the identity (since unitary)
+        assert np.allclose((x ** 0).v, Identity(d).v)
+
+        # check that it's still unitary after fractional binding
+        sqrt_x = x ** 0.5
+        x_check = sqrt_x * sqrt_x
+
+        assert np.allclose(np.linalg.norm(sqrt_x.v), 1)
+        assert np.allclose(np.linalg.norm(x_check.v), 1)
+        assert np.allclose(x_check.v, x.v)
+
+
+def test_degenerate():
+    # tests that an error occurs if throwing away imag components
+    # this error occurs regardless of whether it's unitary
+    x = SemanticPointer([1, 2, 3, 4, 5, 6]).unitary()
+    with pytest.raises(ValueError, match="nondegenerate"):
+        x ** 0.5
+
+
+def test_fractional_binding_natural():
+    # tests fractional binding with natural exponents
+    x = SemanticPointer([1, 2, 3, 4, 5, 6])
+
+    assert np.allclose((x ** 1).v, x.v)
+    assert np.allclose((x ** 2).v, (x * x).v)
+    assert np.allclose((x ** 4).v, (x * x * x * x).v)
+
+    # check that x is degenerate (i.e., not nondegenerate)
+    assert not np.allclose(x.nondegenerate().v, x.v)
+
+
+@pytest.mark.parametrize("d", [1, 4, 9, 144])
+def test_fractional_binding_vtb(d, rng):
+    pytest.importorskip("scipy")
+
+    algebra = VtbAlgebra()
+    v = next(UnitLengthVectors(d, rng=rng))
+    x = SemanticPointer(v, algebra=algebra)
+
+    # note: doesn't quite behave with ideal properties
+    assert np.allclose((x ** 0).v, algebra.identity_element(d))
+    assert np.allclose((x ** 1).v, (~x).v)
+    assert np.allclose((x ** 2).v, (~x * x).v)
+    assert np.allclose((x ** 3).v, (((~x) * x) * x).v)
+
+
+@pytest.mark.parametrize("d", [1, 4, 9, 144])
+def test_nondegenerate_vtb(d, rng):
+    pytest.importorskip("scipy")
+
+    algebra = VtbAlgebra()
+    identity = Identity(d, algebra=algebra)
+    v = next(UnitLengthVectors(d, rng=rng))
+    x = SemanticPointer(v, algebra=algebra).nondegenerate()
+
+    # nondegenerate VTB vectors are unitary and their own inverse
+    assert np.allclose((~x).v, x.v)
+    assert np.allclose(algebra.make_unitary(x.v), x.v)
+
+    # nondegenerate properties are more consistent with HRRs, but they always
+    # oscillate with a period of 2 because V^2 = I
```

Comment:
I'm curious why this happens, is this due to using a Householder matrix for the nondegenerate case? From my experiments using fractional VTB binding I got it to work with just unitary vectors, though it required a slight redefinition of the binding operator to make it commutative. I think this other fractional binding version may be useful to include as well, to allow unique powers greater than 2 and to not require negative powers to be equivalent to the positive ones.

Reply:
Yeah this is just what happens when using the householder matrix to generate those vectors. I didn't see another solution. Would you be open to submitting your version of VTB fractional binding to overwrite mine (and updating the unit tests to match their properties accordingly)?

Reply:
While making the tests I realized my version doesn't actually work for fractional powers for all seeds of unitary vectors, some have the dot product decay after repeated binds. Interestingly, it seems to work for exactly 50% of them (I got lucky/unlucky that everything I tried in the past worked). 100% of them work for repeated binds of integer powers. Definitely would be useful to characterize what makes them work and have a make_nondegenerate() based on that.

The binding change was simply to always bind with the inverse, since binding seemed to act more like subtraction, and 'flipping the sign' made it become more like addition and commutative. If you start by always binding on to the identity, you don't actually need this change.

I'll make a PR with the tests for now, haven't got around to a new make_nondegenerate() yet

Reply:
I have to think more about this degenerate vectors stuff and understand it better, but this statement seem weird to me:

> If you start by always binding on to the identity, you don't actually need this change.

How could binding to the identity have any effect? Isn't the definition of the identity, that it does not change the bound SP?

Reply:
VTB is a little weird like that:
```
vocab = nengo_spa.Vocabulary(64, algebra=nengo_spa.algebras.VtbAlgebra())
vocab.populate('X')

assert np.allclose((vocab['Identity'] * vocab['X']).v, (~vocab['X']).v)
```
Binding with the identity on the LHS will invert (transpose/permute) the RHS.

Reply:
> How could binding to the identity have any effect? Isn't the definition of the identity, that it does not change the bound SP?

In this case the identity is just a right-identity, and not a true identity. A\*I = A, but I\*A != A in general. Correct me if I'm wrong, but I don't think there is a true identity in the VTB algebra. In this case I\*A = A^(-1). By 'binding on to the identity' I meant the second case, where you always start with the right-identity on the left side, and bind SPs from there. This trick allows repeated bindings with fractional powers as defined in `fractional_bind` to behave with similar properties as the HRR version, by flipping the first element bound to align with future bindings.

Reply:
This ^ might also be an apt summary of the difference between how I was going about it in this PR versus how Brent is going about it in #247. This is also why I had the unit test `assert np.allclose((x ** 1).v, (~x).v)`.

Reply:
The identity in VTB is just a right sided identity. Potentially this should be made clearer? Though it would probably involve some breaking changes to the API ... There should also be a left-identity (which should be produced by applying the swapping matrix to the right identity, I believe)

Reply:
I realized that with the current definition of binding in the VTB, there is actually no left-identity. Due to the orientation of the `V_y'` matrix there can be no vector `x` that selects the right elements from `V_y'`, so that `V_y * x = y`. This can be fixed by transposing `V_y` as @bjkomer also pointed out in #247. I think this change would be desirable as it seems to give nicer mathematical properties. Though, we need should verify that this does not break properties in other places.

Reply:
I verified that the plots from the original paper turn out the same even with the transpose. Still have to verify what the impact on the mathematical properties is (we already know that it results in a left-identity, but what about other things, e.g. does it influence commutativity?)

Reply:
That's good to hear! I've played around with it a bit experimentally, and from what I've seen things are still non-commutative (except interestingly in 1D and 4D it is commutative for non-degenerate vectors, not true above that). Possibly there is a different higher dimensional subspace that is commutative, but "unitary with positive determinant" seems to be the only restriction that SSPs need for their properties.

Reply:
Commutativity and associativity properties are not changed by the transpose. However, I found one difference: there isn't a swapping matrix anymore, i.e. it is not possible to swap the operands in bound state.

To me it is not obvious whether one variant should be preferred:
* Original implementation which has a swapping matrix, but no left-identity (and probably no left-inverse, but I didn't verify this yet)
* Implementation with transposed binding matrix, which has not swapping matrix, but a left-identity, in fact left-identity = right-identity (I believe) and thus allows for proper fractional binding powers. (What about the left-inverse?).

Does any of you think that either implementation is strictly better? Otherwise, the best might be to add a third algebra as a variant of the existing VTB. It might make sense to have the original VTB not support fractional binding powers. Furthermore, the API should probably be changed to clarify what is a left or right identity (or both), though this will be a breaking change.

Reply:
Actually, right now the implementation converts the matrix (obtained from the fractional matrix power) by multiplying with the left-identity (which is wrong). Wouldn't it be possible to take the matrix and reshape it back into vector?